File: src/test/java/com/gu/conf/ProjectedConfigurationTest.java
/*
 * Copyright 2010 Guardian News and Media
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.gu.conf;

import com.gu.conf.exceptions.PropertyNotSetException;
import org.junit.Before;
import org.junit.Test;

import java.util.HashSet;
import java.util.Set;

import static java.util.Arrays.asList;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

public class ProjectedConfigurationTest extends ConfigurationAdaptorTestBase {

   @Before
   public void setUp() {
      MapBasedConfiguration mapBasedConfiguration = new MapBasedConfiguration("test");

      mapBasedConfiguration.add("double.property", "25.0");
      mapBasedConfiguration.add("precendence.test.property", "second");
      mapBasedConfiguration.add("integer.property", "23");
      mapBasedConfiguration.add("nonnumeric.property", "qwe");
      mapBasedConfiguration.add("list.property", "rimbaud,verlaine");

      mapBasedConfiguration.add("projected.out", "lost");

      Set<String> projectionNames = new HashSet<String>(asList(
         "precendence.test.property",
         "double.property",
         "integer.property",
         "nonnumeric.property",
         "list.property")
      );

      configuration = new ProjectedConfiguration(mapBasedConfiguration, projectionNames);
   }

   @Test
   public void shouldContainedProjectedProperties() {
      assertThat(configuration.size(), is(5));
      assertThat(configuration.hasProperty("precendence.test.property"), is(true));
      assertThat(configuration.hasProperty("double.property"), is(true));
      assertThat(configuration.hasProperty("integer.property"), is(true));
      assertThat(configuration.hasProperty("nonnumeric.property"), is(true));
      assertThat(configuration.hasProperty("list.property"), is(true));
   }

   @Test
   public void shouldNotContainedProjectedOutProperties() {
      assertThat(configuration.hasProperty("projected.out"), is(false));
   }

   @Test
   public void shouldHaveSamePropertyValuesAsOriginal() throws PropertyNotSetException {
      assertThat(configuration.getStringProperty("precendence.test.property"), is("second"));
      assertThat(configuration.getStringProperty("double.property"), is("25.0"));
      assertThat(configuration.getStringProperty("integer.property"), is("23"));
      assertThat(configuration.getStringProperty("nonnumeric.property"), is("qwe"));
      assertThat(configuration.getStringProperty("list.property"), is("rimbaud,verlaine"));
   }

   @Test
   public void shouldGetPropertySource() {
      String propertySource = configuration.getPropertySource("nonnumeric.property");
      assertThat(propertySource, is("test"));
   }

   @Test
   public void shouldToStringInStandardFormat() {
      String expected =
         "# Properties from " + configuration.getIdentifier() + "\n" +
         "double.property=25.0\n" +
         "integer.property=23\n" +
         "list.property=rimbaud,verlaine\n" +
         "nonnumeric.property=qwe\n" +
         "precendence.test.property=second\n" +
         "\n";

      assertThat(configuration.toString(), is(expected));
   }
}

File: src/test/scala/com/gu/conf/impl/ProjectedConfigurationTest.scala

/*
 * Copyright 2010 Guardian News and Media
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package com.gu.conf.impl

import com.gu.conf.Configuration
import org.scalatest.BeforeAndAfter

class ProjectedConfigurationTest extends AbstractConfigurationTestBase with BeforeAndAfter {
  var original: Configuration = _

  before {
    val mapBasedConfiguration = new MapBasedConfiguration("test")
    mapBasedConfiguration.add("double.property", "25.0")
    mapBasedConfiguration.add("precendence.test.property", "second")
    mapBasedConfiguration.add("integer.property", "23")
    mapBasedConfiguration.add("nonnumeric.property", "qwe")
    mapBasedConfiguration.add("list.property", "rimbaud,verlaine")
    mapBasedConfiguration.add("utility.property", "utility")
    mapBasedConfiguration.add("projected.out", "lost")

    val projectionNames = Set(
      "precendence.test.property",
      "double.property",
      "integer.property",
      "nonnumeric.property",
      "list.property",
      "utility.property")

    configuration = new ProjectedConfiguration(mapBasedConfiguration, projectionNames)
    original = mapBasedConfiguration
  }

  test("should contain projected properties") {
    configuration.size() should be(6)
    configuration.hasProperty("precendence.test.property") should be(true)
    configuration.hasProperty("double.property") should be(true)
    configuration.hasProperty("integer.property") should be(true)
    configuration.hasProperty("nonnumeric.property") should be(true)
    configuration.hasProperty("list.property") should be(true)
    configuration.hasProperty("utility.property") should be(true)
  }

  test("should not contain projected out properties") {
    configuration.hasProperty("projected.out") should be(false)
  }

  test("should have same property values as original") {
    configuration("precendence.test.property") should be("second")
    configuration("double.property") should be("25.0")
    configuration("integer.property") should be("23")
    configuration("nonnumeric.property") should be("qwe")
    configuration("list.property") should be("rimbaud,verlaine")
  }

  test("should get property source") {
    configuration.getPropertySource("nonnumeric.property").get should be(original)
  }

  test("should toString() in standard format") {
    configuration.toString() should be(
      "# Properties from " + configuration.getIdentifier + "\n" +
        "double.property=25.0\n" +
        "integer.property=23\n" +
        "list.property=rimbaud,verlaine\n" +
        "nonnumeric.property=qwe\n" +
        "precendence.test.property=second\n" +
        "utility.property=utility\n" +
        "\n")
  }
}